src/shogun/classifier/NearestCentroid.cpp

```diff
@@ -54,51 +53,41 @@ namespace shogun{
 		{
 			data = distance->get_lhs();
 		}
-		int32_t num_vectors = data->get_num_vectors();
-		int32_t num_classes = ((CMulticlassLabels*) m_labels)->get_num_classes();
-		int32_t num_feats = ((CDenseFeatures<float64_t>*) data)->get_num_features();
-		SGMatrix<float64_t> centroids(num_feats,num_classes);
-		centroids.zero();
 
-		int64_t* num_per_class = new int64_t[num_classes];
-		for (int32_t i=0 ; i<num_classes ; i++)
-		{
-			num_per_class[i]=0;
-		}
+		auto* multiclass_labels = m_labels->as<CMulticlassLabels>();
+		auto* dense_data = data->as<CDenseFeatures<float64_t>>();
+
+		int32_t num_classes = multiclass_labels->get_num_classes();
+		int32_t num_feats = dense_data->get_num_features();
 
-		for (int32_t idx=0 ; idx<num_vectors ; idx++)
+		SGMatrix<float64_t> centroids(num_feats, num_classes);
+		SGVector<int64_t> num_per_class(num_classes);
+
+		auto iter_labels = multiclass_labels->get_labels().begin();
+		for (const auto& current : DotIterator(dense_data))
```

Comment:
we should implement a zip-like iterator for iterating through labels and features at the same time...at some point